## Setting up the files and getting an overview of its information

### Setup

In [92]:
import pandas as pd

In [93]:
data = pd.read_csv('../Data_prep/Louisville_Animal.csv')

In [94]:
data.head()

kennel animalid jurisdiction intype insubtype               indate  \
0  FOLLOWUP  A713651        40202  STRAY       OTC  2022/04/02 00:00:00   
1       N46  A713652        40208  STRAY       OTC  2022/04/02 00:00:00   
2       N17  A733898        40208  STRAY       OTC  2023/03/17 00:00:00   
3       N30  A712650        40222  STRAY       OTC  2022/03/16 00:00:00   
4       409  A743389        40216  STRAY       OTC  2023/08/14 00:00:00   

  surreason   outtype  outsubtype              outdate animaltype sex bites  \
0     STRAY       RTO   IN KENNEL  2022/04/05 00:00:00        DOG   F     N   
1     STRAY  ADOPTION     WALK IN  2022/04/07 00:00:00        CAT   N     N   
2     STRAY    FOSTER    VACATION  2023/04/12 00:00:00        DOG   S     N   
3     STRAY  ADOPTION    INTERNET  2022/03/29 00:00:00        DOG   N     Y   
4     STRAY      EUTH  BEHAV HIST  2023/08/23 00:00:00        DOG   M     Y   

  petsize              color              breed sourcezipcode  ObjectId  
0     MED        TAN / BLACK  ENG BULLDOG / MIX         40208         1  
1   LARGE          ORG TABBY        DOMESTIC SH         40208         2  
2     MED  CHOCOLATE / WHITE           PIT BULL         40208         3  
3   PUPPY                TAN     PIT BULL / MIX         40208         4  
4     MED         BL BRINDLE           PIT BULL         40216         5

In [95]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52606 entries, 0 to 52605
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   kennel         52606 non-null  object
 1   animalid       52606 non-null  object
 2   jurisdiction   40356 non-null  object
 3   intype         52606 non-null  object
 4   insubtype      52549 non-null  object
 5   indate         52606 non-null  object
 6   surreason      40356 non-null  object
 7   outtype        41974 non-null  object
 8   outsubtype     32056 non-null  object
 9   outdate        42002 non-null  object
 10  animaltype     52606 non-null  object
 11  sex            51732 non-null  object
 12  bites          40362 non-null  object
 13  petsize        50333 non-null  object
 14  color          52604 non-null  object
 15  breed          52579 non-null  object
 16  sourcezipcode  46975 non-null  object
 17  ObjectId       52606 non-null  int64 
dtypes: int64(1), object(17)
me

### Cleaning

#### Checking for duplicates

In [96]:
#drop duplicates
data.drop_duplicates(inplace=True)

No duplicates

In [97]:
data.nunique()

kennel             501
animalid         43736
jurisdiction        37
intype              10
insubtype           28
indate            1699
surreason           60
outtype             15
outsubtype          64
outdate           1724
animaltype           7
sex                  5
bites                2
petsize              7
color              468
breed             1231
sourcezipcode      528
ObjectId         52606
dtype: int64

Notes from nunique -- 501 kennels, 43,736 unique animals (indicating that some appear mutliple times), 37 jurisdictions, ten types of intypes, with 28 subtypes, only 15 types of our with 64 out subtypes (may choose to only classify to the 15 for first iteration.)
Currently are 1231 breeds. may be worth looking at grouping/simplifying those. Also may be worth it to clean the animal type to just a few. 
5 sex types?

Subnotes after review of codebook: start with in and out types, then go from there but surrtypes is very important to keep specific. 5 sexes are male, female, spayed, neutured, and unknown. Spayed = female, neutured = male

In [98]:
#check info about animals in multiple times
data[data['animalid'].duplicated(keep=False)].sort_values(by='animalid')

kennel animalid jurisdiction      intype   insubtype  \
46298     N04  A332960        40209       STRAY       FIELD   
46922     301  A332960        40213       STRAY       FIELD   
49245  FOSTER  A341716        40216  CONFISCATE  OWNER DIED   
25479      NP  A341716        40223      FOSTER      RETURN   
19209     238  A358112        40213      FOSTER      RETURN   
...       ...      ...          ...         ...         ...   
4152   FOSTER  A743783        40218       STRAY         OTC   
3830      NF1  A743784        40218      FOSTER      RETURN   
3826   FOSTER  A743784        40218       STRAY         OTC   
4426      NF1  A743785        40218      FOSTER      RETURN   
4751   FOSTER  A743785        40218       STRAY         OTC   

                    indate   surreason   outtype  outsubtype  \
46298  2019/09/21 00:00:00       STRAY       RTO   IN KENNEL   
46922  2020/02/07 00:00:00       STRAY      EUTH   REQUESTED   
49245  2020/08/25 00:00:00  OWNER DIED    FOSTER     TOO OLD   
25479  2020/09/04 00:00:00  FOSTER RET  ADOPTION      FOSTER   
19209  2019/08/07 00:00:00       STRAY  TRANSFER  RESCUE GRP   
...                    ...         ...       ...         ...   
4152   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
3830   2023/08/24 00:00:00       STRAY       NaN         NaN   
3826   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
4426   2023/08/24 00:00:00       STRAY  ADOPTION    INTERNET   
4751   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   

                   outdate animaltype sex bites petsize               color  \
46298  2019/10/25 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
46922  2020/02/11 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
49245  2020/08/31 00:00:00        DOG   S     N     TOY                 RED   
25479  2020/09/04 00:00:00        DOG   S     N     TOY                 RED   
19209  2019/08/07 00:00:00        DOG   S     N     MED       BLACK / BLACK   
...                    ...        ...  ..   ...     ...                 ...   
4152   2023/08/21 00:00:00        CAT   N     N   KITTN  WHITE / GRAY TABBY   
3830                   NaN        CAT   N     N   KITTN          GRAY TABBY   
3826   2023/08/21 00:00:00        CAT   N     N   KITTN          GRAY TABBY   
4426   2023/08/26 00:00:00        CAT   S     N   KITTN                GRAY   
4751   2023/08/21 00:00:00        CAT   S     N   KITTN                GRAY   

                    breed sourcezipcode  ObjectId  
46298  AMER BULLDOG / MIX         40218     46299  
46922  AMER BULLDOG / MIX         40215     46923  
49245          POMERANIAN         40223     49246  
25479          POMERANIAN         40223     25480  
19209       LABRADOR RETR         40218     19210  
...                   ...           ...       ...  
4152          DOMESTIC SH         40218      4153  
3830          DOMESTIC SH         40218      3831  
3826          DOMESTIC SH         40218      3827  
4426          DOMESTIC SH         40218      4427  
4751          DOMESTIC SH         40218      4752  

[15907 rows x 18 columns]

In [99]:
#drop animals with no outtype
data.dropna(subset=['outtype'], inplace=True)

In [100]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41974 entries, 0 to 52605
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   kennel         41974 non-null  object
 1   animalid       41974 non-null  object
 2   jurisdiction   40177 non-null  object
 3   intype         41974 non-null  object
 4   insubtype      41918 non-null  object
 5   indate         41974 non-null  object
 6   surreason      40177 non-null  object
 7   outtype        41974 non-null  object
 8   outsubtype     32056 non-null  object
 9   outdate        41965 non-null  object
 10  animaltype     41974 non-null  object
 11  sex            41829 non-null  object
 12  bites          40177 non-null  object
 13  petsize        41572 non-null  object
 14  color          41972 non-null  object
 15  breed          41974 non-null  object
 16  sourcezipcode  41696 non-null  object
 17  ObjectId       41974 non-null  int64 
dtypes: int64(1), object(17)
me

In [101]:
#count how many times animals that appear multiple times appear
data['animalid'].value_counts().value_counts().sort_values(ascending=False)

1     26151
2      5613
3      1098
4       196
5        64
6        18
7         9
10        2
8         1
Name: animalid, dtype: int64

#### Creating new relevant columns

In [102]:
data

kennel animalid jurisdiction      intype insubtype  \
0      FOLLOWUP  A713651        40202       STRAY       OTC   
1           N46  A713652        40208       STRAY       OTC   
2           N17  A733898        40208       STRAY       OTC   
3           N30  A712650        40222       STRAY       OTC   
4           409  A743389        40216       STRAY       OTC   
...         ...      ...          ...         ...       ...   
52601       N18  A671399        40219       STRAY       OTC   
52602       306  A676059        40211       STRAY     FIELD   
52603        NP  A676364        40215  CONFISCATE   NEGLECT   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT   
52605       VET  A671364        40211       STRAY       OTC   

                    indate  surreason   outtype  outsubtype  \
0      2022/04/02 00:00:00      STRAY       RTO   IN KENNEL   
1      2022/04/02 00:00:00      STRAY  ADOPTION     WALK IN   
2      2023/03/17 00:00:00      STRAY    FOSTER    VACATION   
3      2022/03/16 00:00:00      STRAY  ADOPTION    INTERNET   
4      2023/08/14 00:00:00      STRAY      EUTH  BEHAV HIST   
...                    ...        ...       ...         ...   
52601  2019/09/21 00:00:00      STRAY  ADOPTION    INTERNET   
52602  2020/01/15 00:00:00      STRAY  TRANSFER         KHS   
52603  2020/01/22 00:00:00  VIOLATION  ADOPTION       EVENT   
52604  2019/06/23 00:00:00    CRUELTY       RTO   IN KENNEL   
52605  2019/09/20 00:00:00      STRAY       RTO   IN KENNEL   

                   outdate animaltype sex bites petsize              color  \
0      2022/04/05 00:00:00        DOG   F     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00        CAT   N     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00        DOG   S     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY                TAN   
4      2023/08/23 00:00:00        DOG   M     Y     MED         BL BRINDLE   
...                    ...        ...  ..   ...     ...                ...   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00        DOG   M     N     MED              BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00        DOG   M     N     MED               BLUE   
52605  2019/09/24 00:00:00        DOG   N     N     MED         BR BRINDLE   

                     breed sourcezipcode  ObjectId  
0        ENG BULLDOG / MIX         40208         1  
1              DOMESTIC SH         40208         2  
2                 PIT BULL         40208         3  
3           PIT BULL / MIX         40208         4  
4                 PIT BULL         40216         5  
...                    ...           ...       ...  
52601  LABRADOR RETR / MIX         40202     52602  
52602             PIT BULL         40202     52603  
52603  BORDER COLLIE / MIX         40202     52604  
52604             PIT BULL         40202     52605  
52605             PIT BULL         40202     52606  

[41974 rows x 18 columns]

In [103]:
#if the breed ends with " / MIX" then remove the " / MIX" and in a new column called "Mixed" put true
data['Mixed'] = data['breed'].str.endswith(' / MIX')
data['breed'] = data['breed'].str.replace(' / MIX', '')
data

kennel animalid jurisdiction      intype insubtype  \
0      FOLLOWUP  A713651        40202       STRAY       OTC   
1           N46  A713652        40208       STRAY       OTC   
2           N17  A733898        40208       STRAY       OTC   
3           N30  A712650        40222       STRAY       OTC   
4           409  A743389        40216       STRAY       OTC   
...         ...      ...          ...         ...       ...   
52601       N18  A671399        40219       STRAY       OTC   
52602       306  A676059        40211       STRAY     FIELD   
52603        NP  A676364        40215  CONFISCATE   NEGLECT   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT   
52605       VET  A671364        40211       STRAY       OTC   

                    indate  surreason   outtype  outsubtype  \
0      2022/04/02 00:00:00      STRAY       RTO   IN KENNEL   
1      2022/04/02 00:00:00      STRAY  ADOPTION     WALK IN   
2      2023/03/17 00:00:00      STRAY    FOSTER    VACATION   
3      2022/03/16 00:00:00      STRAY  ADOPTION    INTERNET   
4      2023/08/14 00:00:00      STRAY      EUTH  BEHAV HIST   
...                    ...        ...       ...         ...   
52601  2019/09/21 00:00:00      STRAY  ADOPTION    INTERNET   
52602  2020/01/15 00:00:00      STRAY  TRANSFER         KHS   
52603  2020/01/22 00:00:00  VIOLATION  ADOPTION       EVENT   
52604  2019/06/23 00:00:00    CRUELTY       RTO   IN KENNEL   
52605  2019/09/20 00:00:00      STRAY       RTO   IN KENNEL   

                   outdate animaltype sex bites petsize              color  \
0      2022/04/05 00:00:00        DOG   F     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00        CAT   N     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00        DOG   S     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY                TAN   
4      2023/08/23 00:00:00        DOG   M     Y     MED         BL BRINDLE   
...                    ...        ...  ..   ...     ...                ...   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00        DOG   M     N     MED              BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00        DOG   M     N     MED               BLUE   
52605  2019/09/24 00:00:00        DOG   N     N     MED         BR BRINDLE   

               breed sourcezipcode  ObjectId  Mixed  
0        ENG BULLDOG         40208         1   True  
1        DOMESTIC SH         40208         2  False  
2           PIT BULL         40208         3  False  
3           PIT BULL         40208         4   True  
4           PIT BULL         40216         5  False  
...              ...           ...       ...    ...  
52601  LABRADOR RETR         40202     52602   True  
52602       PIT BULL         40202     52603  False  
52603  BORDER COLLIE         40202     52604   True  
52604       PIT BULL         40202     52605  False  
52605       PIT BULL         40202     52606  False  

[41974 rows x 19 columns]

In [104]:
#create a new column called time_in that is indate - outdate
data['days_in'] = pd.to_datetime(data['outdate']) - pd.to_datetime(data['indate'])
data['days_in'] = data['days_in'].dt.days
data

kennel animalid jurisdiction      intype insubtype  \
0      FOLLOWUP  A713651        40202       STRAY       OTC   
1           N46  A713652        40208       STRAY       OTC   
2           N17  A733898        40208       STRAY       OTC   
3           N30  A712650        40222       STRAY       OTC   
4           409  A743389        40216       STRAY       OTC   
...         ...      ...          ...         ...       ...   
52601       N18  A671399        40219       STRAY       OTC   
52602       306  A676059        40211       STRAY     FIELD   
52603        NP  A676364        40215  CONFISCATE   NEGLECT   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT   
52605       VET  A671364        40211       STRAY       OTC   

                    indate  surreason   outtype  outsubtype  \
0      2022/04/02 00:00:00      STRAY       RTO   IN KENNEL   
1      2022/04/02 00:00:00      STRAY  ADOPTION     WALK IN   
2      2023/03/17 00:00:00      STRAY    FOSTER    VACATION   
3      2022/03/16 00:00:00      STRAY  ADOPTION    INTERNET   
4      2023/08/14 00:00:00      STRAY      EUTH  BEHAV HIST   
...                    ...        ...       ...         ...   
52601  2019/09/21 00:00:00      STRAY  ADOPTION    INTERNET   
52602  2020/01/15 00:00:00      STRAY  TRANSFER         KHS   
52603  2020/01/22 00:00:00  VIOLATION  ADOPTION       EVENT   
52604  2019/06/23 00:00:00    CRUELTY       RTO   IN KENNEL   
52605  2019/09/20 00:00:00      STRAY       RTO   IN KENNEL   

                   outdate animaltype sex bites petsize              color  \
0      2022/04/05 00:00:00        DOG   F     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00        CAT   N     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00        DOG   S     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00        DOG   N     Y   PUPPY                TAN   
4      2023/08/23 00:00:00        DOG   M     Y     MED         BL BRINDLE   
...                    ...        ...  ..   ...     ...                ...   
52601  2019/09/28 00:00:00        DOG   N     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00        DOG   M     N     MED              BROWN   
52603  2020/02/15 00:00:00        DOG   S     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00        DOG   M     N     MED               BLUE   
52605  2019/09/24 00:00:00        DOG   N     N     MED         BR BRINDLE   

               breed sourcezipcode  ObjectId  Mixed  days_in  
0        ENG BULLDOG         40208         1   True      3.0  
1        DOMESTIC SH         40208         2  False      5.0  
2           PIT BULL         40208         3  False     26.0  
3           PIT BULL         40208         4   True     13.0  
4           PIT BULL         40216         5  False      9.0  
...              ...           ...       ...    ...      ...  
52601  LABRADOR RETR         40202     52602   True      7.0  
52602       PIT BULL         40202     52603  False      9.0  
52603  BORDER COLLIE         40202     52604   True     24.0  
52604       PIT BULL         40202     52605  False      4.0  
52605       PIT BULL         40202     52606  False      4.0  

[41974 rows x 20 columns]

In [105]:
#create two new columns in_season and out_season that are fall, winter, spring, summer based on the month of indate and outdate
data['in_season'] = pd.to_datetime(data['indate']).dt.month.map({1: 'WINTER', 2: 'WINTER', 3: 'SPRING', 4: 'SPRING', 5: 'SPRING', 6: 'SUMMER', 7: 'SUMMER', 8: 'SUMMER', 9: 'FALL', 10: 'FALL', 11: 'FALL', 12: 'WINTER'})
data['out_season'] = pd.to_datetime(data['outdate']).dt.month.map({1: 'WINTER', 2: 'WINTER', 3: 'SPRING', 4: 'SPRING', 5: 'SPRING', 6: 'SUMMER', 7: 'SUMMER', 8: 'SUMMER', 9: 'FALL', 10: 'FALL', 11: 'FALL', 12: 'WINTER'})
data

kennel animalid jurisdiction      intype insubtype  \
0      FOLLOWUP  A713651        40202       STRAY       OTC   
1           N46  A713652        40208       STRAY       OTC   
2           N17  A733898        40208       STRAY       OTC   
3           N30  A712650        40222       STRAY       OTC   
4           409  A743389        40216       STRAY       OTC   
...         ...      ...          ...         ...       ...   
52601       N18  A671399        40219       STRAY       OTC   
52602       306  A676059        40211       STRAY     FIELD   
52603        NP  A676364        40215  CONFISCATE   NEGLECT   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT   
52605       VET  A671364        40211       STRAY       OTC   

                    indate  surreason   outtype  outsubtype  \
0      2022/04/02 00:00:00      STRAY       RTO   IN KENNEL   
1      2022/04/02 00:00:00      STRAY  ADOPTION     WALK IN   
2      2023/03/17 00:00:00      STRAY    FOSTER    VACATION   
3      2022/03/16 00:00:00      STRAY  ADOPTION    INTERNET   
4      2023/08/14 00:00:00      STRAY      EUTH  BEHAV HIST   
...                    ...        ...       ...         ...   
52601  2019/09/21 00:00:00      STRAY  ADOPTION    INTERNET   
52602  2020/01/15 00:00:00      STRAY  TRANSFER         KHS   
52603  2020/01/22 00:00:00  VIOLATION  ADOPTION       EVENT   
52604  2019/06/23 00:00:00    CRUELTY       RTO   IN KENNEL   
52605  2019/09/20 00:00:00      STRAY       RTO   IN KENNEL   

                   outdate  ... bites petsize              color  \
0      2022/04/05 00:00:00  ...     N     MED        TAN / BLACK   
1      2022/04/07 00:00:00  ...     N   LARGE          ORG TABBY   
2      2023/04/12 00:00:00  ...     N     MED  CHOCOLATE / WHITE   
3      2022/03/29 00:00:00  ...     Y   PUPPY                TAN   
4      2023/08/23 00:00:00  ...     Y     MED         BL BRINDLE   
...                    ...  ...   ...     ...                ...   
52601  2019/09/28 00:00:00  ...     N   LARGE      BLACK / BROWN   
52602  2020/01/24 00:00:00  ...     N     MED              BROWN   
52603  2020/02/15 00:00:00  ...     N     MED      BLACK / WHITE   
52604  2019/06/27 00:00:00  ...     N     MED               BLUE   
52605  2019/09/24 00:00:00  ...     N     MED         BR BRINDLE   

               breed sourcezipcode ObjectId  Mixed  days_in  in_season  \
0        ENG BULLDOG         40208        1   True      3.0     SPRING   
1        DOMESTIC SH         40208        2  False      5.0     SPRING   
2           PIT BULL         40208        3  False     26.0     SPRING   
3           PIT BULL         40208        4   True     13.0     SPRING   
4           PIT BULL         40216        5  False      9.0     SUMMER   
...              ...           ...      ...    ...      ...        ...   
52601  LABRADOR RETR         40202    52602   True      7.0       FALL   
52602       PIT BULL         40202    52603  False      9.0     WINTER   
52603  BORDER COLLIE         40202    52604   True     24.0     WINTER   
52604       PIT BULL         40202    52605  False      4.0     SUMMER   
52605       PIT BULL         40202    52606  False      4.0       FALL   

       out_season  
0          SPRING  
1          SPRING  
2          SPRING  
3          SPRING  
4          SUMMER  
...           ...  
52601        FALL  
52602      WINTER  
52603      WINTER  
52604      SUMMER  
52605        FALL  

[41974 rows x 22 columns]

In [106]:
#drop indate and outdate columns
data.drop(columns=['indate', 'outdate'], inplace=True)

In [107]:
#for breeds with multiple breeds, split the breeds into two columns and update mixed column to true if there were false before. If mixed is true, but breed2 is NaN, update to "unknown"
data['breed'] = data['breed'].str.replace(' / ', ',')
data['breed1'] = data['breed'].str.split(',').str[0]
data['breed2'] = data['breed'].str.split(',').str[1]
data['Mixed'] = data['Mixed'] | data['breed'].str.contains(',')
data

kennel animalid jurisdiction      intype insubtype  surreason  \
0      FOLLOWUP  A713651        40202       STRAY       OTC      STRAY   
1           N46  A713652        40208       STRAY       OTC      STRAY   
2           N17  A733898        40208       STRAY       OTC      STRAY   
3           N30  A712650        40222       STRAY       OTC      STRAY   
4           409  A743389        40216       STRAY       OTC      STRAY   
...         ...      ...          ...         ...       ...        ...   
52601       N18  A671399        40219       STRAY       OTC      STRAY   
52602       306  A676059        40211       STRAY     FIELD      STRAY   
52603        NP  A676364        40215  CONFISCATE   NEGLECT  VIOLATION   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT    CRUELTY   
52605       VET  A671364        40211       STRAY       OTC      STRAY   

        outtype  outsubtype animaltype sex  ...              color  \
0           RTO   IN KENNEL        DOG   F  ...        TAN / BLACK   
1      ADOPTION     WALK IN        CAT   N  ...          ORG TABBY   
2        FOSTER    VACATION        DOG   S  ...  CHOCOLATE / WHITE   
3      ADOPTION    INTERNET        DOG   N  ...                TAN   
4          EUTH  BEHAV HIST        DOG   M  ...         BL BRINDLE   
...         ...         ...        ...  ..  ...                ...   
52601  ADOPTION    INTERNET        DOG   N  ...      BLACK / BROWN   
52602  TRANSFER         KHS        DOG   M  ...              BROWN   
52603  ADOPTION       EVENT        DOG   S  ...      BLACK / WHITE   
52604       RTO   IN KENNEL        DOG   M  ...               BLUE   
52605       RTO   IN KENNEL        DOG   N  ...         BR BRINDLE   

               breed sourcezipcode ObjectId  Mixed  days_in  in_season  \
0        ENG BULLDOG         40208        1   True      3.0     SPRING   
1        DOMESTIC SH         40208        2  False      5.0     SPRING   
2           PIT BULL         40208        3  False     26.0     SPRING   
3           PIT BULL         40208        4   True     13.0     SPRING   
4           PIT BULL         40216        5  False      9.0     SUMMER   
...              ...           ...      ...    ...      ...        ...   
52601  LABRADOR RETR         40202    52602   True      7.0       FALL   
52602       PIT BULL         40202    52603  False      9.0     WINTER   
52603  BORDER COLLIE         40202    52604   True     24.0     WINTER   
52604       PIT BULL         40202    52605  False      4.0     SUMMER   
52605       PIT BULL         40202    52606  False      4.0       FALL   

       out_season         breed1 breed2  
0          SPRING    ENG BULLDOG    NaN  
1          SPRING    DOMESTIC SH    NaN  
2          SPRING       PIT BULL    NaN  
3          SPRING       PIT BULL    NaN  
4          SUMMER       PIT BULL    NaN  
...           ...            ...    ...  
52601        FALL  LABRADOR RETR    NaN  
52602      WINTER       PIT BULL    NaN  
52603      WINTER  BORDER COLLIE    NaN  
52604      SUMMER       PIT BULL    NaN  
52605        FALL       PIT BULL    NaN  

[41974 rows x 22 columns]

In [108]:
#if mixed is true, but breed2 is NaN, update to "unknown"
data.loc[(data['Mixed'] == True) & (data['breed2'].isna()), 'breed2'] = 'UNKNOWN'
data

kennel animalid jurisdiction      intype insubtype  surreason  \
0      FOLLOWUP  A713651        40202       STRAY       OTC      STRAY   
1           N46  A713652        40208       STRAY       OTC      STRAY   
2           N17  A733898        40208       STRAY       OTC      STRAY   
3           N30  A712650        40222       STRAY       OTC      STRAY   
4           409  A743389        40216       STRAY       OTC      STRAY   
...         ...      ...          ...         ...       ...        ...   
52601       N18  A671399        40219       STRAY       OTC      STRAY   
52602       306  A676059        40211       STRAY     FIELD      STRAY   
52603        NP  A676364        40215  CONFISCATE   NEGLECT  VIOLATION   
52604       Q15  A601911        40216  CONFISCATE   NEGLECT    CRUELTY   
52605       VET  A671364        40211       STRAY       OTC      STRAY   

        outtype  outsubtype animaltype sex  ...              color  \
0           RTO   IN KENNEL        DOG   F  ...        TAN / BLACK   
1      ADOPTION     WALK IN        CAT   N  ...          ORG TABBY   
2        FOSTER    VACATION        DOG   S  ...  CHOCOLATE / WHITE   
3      ADOPTION    INTERNET        DOG   N  ...                TAN   
4          EUTH  BEHAV HIST        DOG   M  ...         BL BRINDLE   
...         ...         ...        ...  ..  ...                ...   
52601  ADOPTION    INTERNET        DOG   N  ...      BLACK / BROWN   
52602  TRANSFER         KHS        DOG   M  ...              BROWN   
52603  ADOPTION       EVENT        DOG   S  ...      BLACK / WHITE   
52604       RTO   IN KENNEL        DOG   M  ...               BLUE   
52605       RTO   IN KENNEL        DOG   N  ...         BR BRINDLE   

               breed sourcezipcode ObjectId  Mixed  days_in  in_season  \
0        ENG BULLDOG         40208        1   True      3.0     SPRING   
1        DOMESTIC SH         40208        2  False      5.0     SPRING   
2           PIT BULL         40208        3  False     26.0     SPRING   
3           PIT BULL         40208        4   True     13.0     SPRING   
4           PIT BULL         40216        5  False      9.0     SUMMER   
...              ...           ...      ...    ...      ...        ...   
52601  LABRADOR RETR         40202    52602   True      7.0       FALL   
52602       PIT BULL         40202    52603  False      9.0     WINTER   
52603  BORDER COLLIE         40202    52604   True     24.0     WINTER   
52604       PIT BULL         40202    52605  False      4.0     SUMMER   
52605       PIT BULL         40202    52606  False      4.0       FALL   

       out_season         breed1   breed2  
0          SPRING    ENG BULLDOG  UNKNOWN  
1          SPRING    DOMESTIC SH      NaN  
2          SPRING       PIT BULL      NaN  
3          SPRING       PIT BULL  UNKNOWN  
4          SUMMER       PIT BULL      NaN  
...           ...            ...      ...  
52601        FALL  LABRADOR RETR  UNKNOWN  
52602      WINTER       PIT BULL      NaN  
52603      WINTER  BORDER COLLIE  UNKNOWN  
52604      SUMMER       PIT BULL      NaN  
52605        FALL       PIT BULL      NaN  

[41974 rows x 22 columns]

In [111]:
#drop breed column
data.drop(columns=['breed'], inplace=True)
#rename breed1 to breed
data.rename(columns={'breed1': 'breed'}, inplace=True)
#renamed breed2 to mixed_breed
data.rename(columns={'breed2': 'mixed_breed'}, inplace=True)

In [112]:
#export to csv
data.to_csv('../Data_prep/Louisville_Animal_Cleaned.csv', index=False)

#### Splitting data into two frames

In [ ]:
#split data into two dataframes: one with animals that appear multiple times and one with animals that appear once
data_multi = data[data['animalid'].duplicated(keep=False)].sort_values(by='animalid')
data_once = data[~data['animalid'].duplicated(keep=False)].sort_values(by='animalid')

In [ ]:
data_multi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15823 entries, 46922 to 4751
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   kennel         15823 non-null  object 
 1   animalid       15823 non-null  object 
 2   jurisdiction   15821 non-null  object 
 3   intype         15823 non-null  object 
 4   insubtype      15796 non-null  object 
 5   indate         15823 non-null  object 
 6   surreason      15821 non-null  object 
 7   outtype        15823 non-null  object 
 8   outsubtype     14129 non-null  object 
 9   outdate        15823 non-null  object 
 10  animaltype     15823 non-null  object 
 11  sex            15823 non-null  object 
 12  bites          15823 non-null  object 
 13  petsize        15823 non-null  object 
 14  color          15821 non-null  object 
 15  breed          15823 non-null  object 
 16  sourcezipcode  15709 non-null  object 
 17  ObjectId       15823 non-null  int64  
 18  Mix

In [ ]:
data_multi.nunique()

kennel             458
animalid          7001
jurisdiction        37
intype               9
insubtype           24
indate            1678
surreason           57
outtype             15
outsubtype          56
outdate           1695
animaltype           7
sex                  5
bites                2
petsize              7
color              256
breed              502
sourcezipcode      214
ObjectId         15823
Mixed                2
days_in            155
dtype: int64

In [ ]:
data_once.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26151 entries, 50277 to 45420
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   kennel         26151 non-null  object 
 1   animalid       26151 non-null  object 
 2   jurisdiction   24356 non-null  object 
 3   intype         26151 non-null  object 
 4   insubtype      26122 non-null  object 
 5   indate         26151 non-null  object 
 6   surreason      24356 non-null  object 
 7   outtype        26151 non-null  object 
 8   outsubtype     17927 non-null  object 
 9   outdate        26142 non-null  object 
 10  animaltype     26151 non-null  object 
 11  sex            26006 non-null  object 
 12  bites          24354 non-null  object 
 13  petsize        25749 non-null  object 
 14  color          26151 non-null  object 
 15  breed          26151 non-null  object 
 16  sourcezipcode  25987 non-null  object 
 17  ObjectId       26151 non-null  int64  
 18  Mi

In [ ]:
data_once.nunique()

kennel             495
animalid         26151
jurisdiction        37
intype              10
insubtype           28
indate            1696
surreason           52
outtype             15
outsubtype          60
outdate           1719
animaltype           7
sex                  5
bites                2
petsize              7
color              421
breed              883
sourcezipcode      305
ObjectId         26151
Mixed                2
days_in            164
dtype: int64

In [ ]:
data_multi

kennel animalid jurisdiction      intype   insubtype  \
46922     301  A332960        40213       STRAY       FIELD   
46298     N04  A332960        40209       STRAY       FIELD   
49245  FOSTER  A341716        40216  CONFISCATE  OWNER DIED   
25479      NP  A341716        40223      FOSTER      RETURN   
46460     238  A358112        40213       STRAY       FIELD   
...       ...      ...          ...         ...         ...   
4027   FOSTER  A743667        40218       STRAY         OTC   
4152   FOSTER  A743783        40218       STRAY         OTC   
4005      NF1  A743783        40218      FOSTER      RETURN   
4426      NF1  A743785        40218      FOSTER      RETURN   
4751   FOSTER  A743785        40218       STRAY         OTC   

                    indate   surreason   outtype  outsubtype  \
46922  2020/02/07 00:00:00       STRAY      EUTH   REQUESTED   
46298  2019/09/21 00:00:00       STRAY       RTO   IN KENNEL   
49245  2020/08/25 00:00:00  OWNER DIED    FOSTER     TOO OLD   
25479  2020/09/04 00:00:00  FOSTER RET  ADOPTION      FOSTER   
46460  2019/06/13 00:00:00       STRAY    FOSTER     TOO OLD   
...                    ...         ...       ...         ...   
4027   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
4152   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   
4005   2023/08/24 00:00:00       STRAY  ADOPTION    INTERNET   
4426   2023/08/24 00:00:00       STRAY  ADOPTION    INTERNET   
4751   2023/08/21 00:00:00       STRAY    FOSTER  TIME/SPACE   

                   outdate animaltype sex bites petsize               color  \
46922  2020/02/11 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
46298  2019/10/25 00:00:00        DOG   N     N   LARGE       BROWN / WHITE   
49245  2020/08/31 00:00:00        DOG   S     N     TOY                 RED   
25479  2020/09/04 00:00:00        DOG   S     N     TOY                 RED   
46460  2019/07/01 00:00:00        DOG   S     N     MED       BLACK / BLACK   
...                    ...        ...  ..   ...     ...                 ...   
4027   2023/08/21 00:00:00        CAT   S     N   KITTN          GRAY TABBY   
4152   2023/08/21 00:00:00        CAT   N     N   KITTN  WHITE / GRAY TABBY   
4005   2023/08/28 00:00:00        CAT   N     N   KITTN  WHITE / GRAY TABBY   
4426   2023/08/26 00:00:00        CAT   S     N   KITTN                GRAY   
4751   2023/08/21 00:00:00        CAT   S     N   KITTN                GRAY   

               breed sourcezipcode  ObjectId  Mixed  days_in  
46922   AMER BULLDOG         40215     46923   True      4.0  
46298   AMER BULLDOG         40218     46299   True     34.0  
49245     POMERANIAN         40223     49246  False      6.0  
25479     POMERANIAN         40223     25480  False      0.0  
46460  LABRADOR RETR         40218     46461  False     18.0  
...              ...           ...       ...    ...      ...  
4027     DOMESTIC SH         40218      4028  False      0.0  
4152     DOMESTIC SH         40218      4153  False      0.0  
4005     DOMESTIC SH         40218      4006  False      4.0  
4426     DOMESTIC SH         40218      4427  False      2.0  
4751     DOMESTIC SH         40218      4752  False      0.0  

[15823 rows x 20 columns]

In [ ]:
#change 0 days_in to N/A
data_multi.loc[data_multi['outdate'] == '0', 'outdate'] = 'N/A'